### DISCUSSION TITLE
Client with a subscription in a "forever loop" and reconnection

### DISCUSSION THREAD

**Discussion by pridi** [FreeOpcUa/opcua-asyncio#623] [Q&A]
I'm quite sure that for many of you will be a very easy question but I'm struggling on this!

What would be the best way to create a subscription to a variable that will never die (a forever loop) and in case of disconnection of the server tries to reconnect and re-establish the connection when the server will return up?
Thank you very much for your help

**Comment by AndreasHeine** [comment]
https://reference.opcfoundation.org/v104/Core/docs/Part4/6.7/

**Reply by pridi** [reply]
Thank you @AndreasHeine. My issue is to create these mechanisms by using the async library. I'm struggling to find out the correct way to have the loop continuously running and relaunch the connection. Do you have an example showing this?

**Reply by pridi** [reply]
This is what I've implemented to try to have the continuous loop...and for sure is not correct! The client reconnection has not been implemented yet...

```python
import asyncio
import logging

from asyncua import Client, Node, ua

logging.basicConfig(level=logging.INFO)
_logger = logging.getLogger('asyncua')


class SubscriptionHandler:
    """
    The SubscriptionHandler is used to handle the data that is received for the subscription.
    """
    def datachange_notification(self, node: Node, val, data):
        """
        Callback for asyncua Subscription.
        This method will be called when the Client received a data change message from the Server.
        """
        print("Python: New data change event", node, val)


async def mySubs():
    client = Client(url="opc.tcp://myip:4840/")
    async with client:
        handler = SubscriptionHandler()
        # We create a Client Subscription.
        subscription = await client.create_subscription(500, handler)

        # We subscribe to data changes for two nodes (variables).
        var_1 = client.get_node("s=myvariabletest")
        await subscription.subscribe_data_change([var_1])

        ## We let the subscription run for 100 seconds
        await asyncio.sleep(100)
        ## We delete the subscription (this un-subscribes from the data changes of the two variables).
        ## This is optional since closing the connection will also delete all subscriptions.
        await subscription.delete()
        ## After one second we exit the Client context manager - this will close the connection.
        await asyncio.sleep(1)


async def main():
    while True:
        await mySubs()


if __name__ == "__main__":
    asyncio.run(main())
```

after 100 sec the function will run again but if the connection with the server will die in the 100sec sleep time, I'm obliged to wait a lot of time to have again the connection

**Reply by AndreasHeine** [reply]
you need a task which reads cyclic the serverstate if it fails cancel the task and create a new

**Comment by jmarshall9120** [comment]
2022 calling.  I'm solving the same problem right now.  Any case you got a working build here @pridi ?  

The subscription doesn't seem to know when the connection fails unfortunately.
